libraries/protocol/include/graphene/protocol/htlc.hpp

```diff
@@ -59,8 +63,13 @@ namespace graphene { namespace protocol {
          uint16_t preimage_size;
          // The time the funds will be returned to the source if not claimed
          uint32_t claim_period_seconds;
-         // for future expansion
-         extensions_type extensions;
+         
+         // additional extensions
+         struct additional_options_type
+         {
+            fc::optional<memo_data> memo;
+         };
```

Comment:
Add the length of memo to the per_kb fee, or a separate one? Opinions?

Reply:
Well, there's a twist. htlc_create_operation has fees for the transaction, plus number of days. It does not have a per_kb fee as I thought. htlc_redeem charges a fee per kb, as it is only at that time that we know the size of the preimage. So we have two options:

1. Create a per_kb fee for htlc_create_operation or
2. Add the memo size to the per_kb fee at redemption, and make the redeemer pay the per_kb fee.

I would say #1 is better. Anyone else have an opinion?

Reply:
I'd recommend go with `1`. However, the implementation would be tricky, because there is no `extensions` field in `htlc_create_operation::fee_parameter_type`.

Reply:
We need to add the fee rate to somewhere in the global properties, then check the paid fee against it in `htlc_create_evaluator`.

Reply:
Implemented option 1. See https://github.com/bitshares/bitshares-core/pull/1998/commits/f1a2ab0778211f3e47068ab1c043395fe8873d74

Reply:
@jmjatlanta please gather information before heading into implementation.

Fee processing is a bit complicated in BitShares.
* fees need to be shared with the referral program
* fees can be in an asset which is not CORE, and there are fee pools
* need to make sure `get_required_fees` API call contains sufficient fee
* adding a field into `fee_parameter_type` is not possible due to serialization changes which affect `committee_member_update_global_parameters_operation`, thus need to add it somewhere else. Due to this, the old `calculate_fee ()` code won't know it, so we need to write new code.

Reply:
I didn't see your comments until after I implemented. Oh well. I'll roll it back. Trying to figure out a way to get calculate_fee() to know about chain parameters may be tricky too. Such a pain. Perhaps versioning in serialization should be considered, but a discussion for another day.

Reply:
What about using the fee_per_kb from transfer_operation? See https://github.com/bitshares/bitshares-core/pull/1998/commits/20e43f0646626a7fd97291fc9c69a886406aef48#diff-f7f633589e27d70fdfaf4fb1a2635aac

Reply:
What about using the transfer_operation's per_kb_fee? Advantage: minimal code change. Disadvantage: can not have a different fee for memos between transfer_operation and htlc_create. See https://github.com/bitshares/bitshares-core/pull/1998/commits/20e43f0646626a7fd97291fc9c69a886406aef48#diff-f7f633589e27d70fdfaf4fb1a2635aac

Reply:
Perhaps doable, but if we can't get it done in a timely manner, I'd recommend that we don't add memo in this release.